### PR TITLE
"Add" support for mDNS

### PR DESCRIPTION
**Pull request by p-l-**
This PR introduces DestIPField and DestIPv6Field, similar to DestMACField, to be used when a default destination address can be guessed based on the payload. The only use I can see for that for now is multicast protocols (such as mDNS), but it might as well be useful in other situations.

### CODE REVIEW

**Comment by guedou** [comment]
Cool patch !

However, I wonder if we could find a more generic way to add this feature using something like overload_fields (used a lot in inet6.py), or a generic hook. This patch could benefit to other layer such as HSRP or OSPF.

**Comment by p-l-** [comment]
Maybe this could be done later, when more protocols are supported?

**Comment by guedou** [comment]
I am afraid that this feature will end up not being used =\

A quick fix could be to move the address/port binding to a dictionary in config, and use a function similar to bind layers in config.

Another option could be to open an issue asking for a generic method to be implemented.

**Comment by p-l-** [comment]
This mechanism works, is efficient, and covers the current needs. Moreover, the code is simple, easy to understand and to extend.

After some attempts, I don't think that a mechanism close to `bind_layers()` would be as efficient. It would also not be as generic: what if the destination address depends on a field value from the layer above the payload? or even on a specific layer which can be several levels above the payload?

For now, I don't have enough time to investigate this further, but I think this code was ready to merge.

**Comment by guedou** [comment]
The last line from my previous comment lost its meaning. Indeed, I meant that
1. I can merge the current PR as it is
2. You can open an enhancement issue asking for a generic method

If you recreate your feature-mdns branch. I might be able to reopen the PR then merge it.

**Comment by p-l-** [comment]
Sorry for the delay, here is an attempts for a generic `DestField` with binding mechanism. What do you think?

File: test/regression.uts [hunk 1]
```diff
@@ -4494,3 +4494,25 @@ assert bpkt.chksum == 0xf0bc and bpkt.payload.chksum == 0xbb17
 pkt = IP(len=42, ihl=6, options=[IPOption_RR()]) / UDP() / ("A" * 10)
 bpkt = IP(str(pkt))
 assert bpkt.chksum == 0xf0bc and bpkt.payload.chksum == 0xbb17
+
+= Layer binding
+
++ Test DestMACField & DestIPField
```
Comment (guedou) [inline test/regression.uts]: For some reason, the HTML output is empty due to '+'. Using '*' instead fixes the issue.

**Comment by p-l-** [comment]
I had totally missed that. Fixed, thanks!

**Comment by Lodags** [comment]
Hello guys, I struggled with mDNS lastly and saw that finally it was not available as a layer. I was motivated to work with this and create a distinct mdns layer, but is it possible to implement this another layer ? And is it relevant ?

**Comment by guedou** [comment]
mDNS being DNS over IP multicast, it works fine out of the box. However I did not use it since a long time.

What did you try?

**Comment by Lodags** [comment]
I wanted to make a Lan scanner with mdns reverse resolution, so I tried this : IP(dst="224.0.0.251")/UDP(dport=5353)/DNS(rd=1,qd=DNSQR(qname=dns,qtype='PTR')). But the answer was empty each time. So I compared the scapy and "dig -x 192.168.50.2 @224.0.0.251 -p 5353 +short" (which works fine) frames with wireshark. The frames were almost identical, except that they dig ones were marked as mdns whereas the scapy's ones, that were marked as dns. So I thought that I had to directly code a mdns layer. (nb : I know the mDNS protocol almost since yesterday)

**Comment by guedou** [comment]
An easy solution is to import the packet sniffed with wireshark in Scapy, using rdpcap(), and compare it to the one that you are building. You can for example use the `hexdiff()` function.

**Comment by Lodags** [comment]
The annoying thing is that Scapy doesn't detect any kind of DNS layer, it just adds its raw layer :
0000 Ether / IP / UDP 192.168.50.43:54904 > 224.0.0.251:mdns / Raw

Here is the Raw (the commentaries describe their above line) : 
        Raw(load='\x03\x83'
                    #Transaction ID
                    '\x01\x00'
                    #Flags
                    '\x00\x01'
                    #Questions
                    '\x00\x00'
                    #Answers RR
                    '\x00\x00'
                    #Authority RR
                    '\x00\x00'
                    #Additionnal RRs
                    '\x02\x32\x39\x02\x00\x10\x35\x30\x03\x31\x36\x38\x03\x31\x39\x32\x07\x69\x6e\x2d\x61\x64\x00\x20\x64\x72\x04\x61\x72\x70\x61\x00'
                    #Name
                    '\x00\x0c'
                    #PT
                    '\x00\x01'))
                    #IN


I tried to make my own by adding these data in the packet and sendp() them, but it doesn't work and wireshark marks it as malformed packet. How would you do it ?

**Comment by guedou** [comment]
The following works fine:
`wireshark(IP(dst="224.0.0.251")/UDP(dport=5353)/DNS(rd=1,qd=DNSQR(qname="192.168.50.2",qtype='PTR')))`

Are you using Scapy from github ?  If this does not fix your problem, please open a new issue on github as PR should not be used to discuss issues. Thanks.

By the way `sendp()` injects at layer 2, and therefore requires an Ether() header before IP()

**Comment by Lodags** [comment]
Thank you, I opened one : https://github.com/secdev/scapy/issues/179

**Pull request merged**